### ISSUE TITLE
Allow flexibility in dmabuf-import use case

### ISSUE THREAD

**Issue by sreerenjb** [Intel-Media-SDK/MediaSDK#155]
Currently, MediaSDK requires to pre-allocate all the surfaces in a pool during MFXXXX_Init().
For dmabuf use case there are two possible options:
1) dmabuf-export from MediaSDK components
2) dmabuf-import in MediaSDK components.

The dmabuf-export(1) works fine since MediaSDK application is creating all surfaces.

But in case of dmabuf-import(2), the Media-SDK application might not be able to preallocate all the dmabuf based memory/surfaces because the original memory has been already created by some other driver (Eg: V4l2) and the msdk-application only get the memory FD of a single buffer just before processing.
Is there any way to bypass the requirement to have to preallocate all required buffers during MFXXXX_Init()?

Note: I'm not saying that it is "impossible" to get all buffer-id before MFX_XXX_Init() within the application, but it would be better to handle the stuff in the other way around (possibility to accept dmabuf based buffers as the input of Encode and VPP as it comes even if it didn't belong to a pre-allocated buffer pool)

**Comment by sreerenjb** [comment]
@artem-shaporenko Good to know that there are some investigations on going!
Thank you.
BTW,  from a middleware perspective,  we are doing some investigations on this in GStreamer side.
https://bugzilla.gnome.org/show_bug.cgi?id=794817

**Comment by rlim77** [comment]
We've implemented a WA in gst-mfx. It's definitely not ideal. The main issue is really iopattern. We need to set midway in the encode/VPP process to indicate it's IN_VIDEO_MEM instead of IN_SYS_MEM. Will like an elegant way to do this.

https://github.com/intel/gstreamer-media-SDK/pull/94/files#diff-3af16745bbf3d57e6dee7290a030a131

**Comment by sreerenjb** [comment]
@rlim77 I'm still thinking whether we should go for a "general approach" by preallocating vasurfaces  from dmabuf exported by other drivers or to replace mem ids dynamically as Artem proposed (which is what you have implemented too).
The so-called "general approach" is a bit weird IMHO.  Maybe it is better to stick with the on-demand surface creation and MemId replacement. Have you tested the dambuf-import feature in your gst-mfx? I believe you have icamerasrc available for testing.right?

**Comment by sreerenjb** [comment]
If we really want to avoid a memcopy in dmabuf-import, it needs a solution for this:
https://github.com/intel/media-driver/issues/169

**Comment by rlim77** [comment]
@sreerenjb Yes. It was tested on KBL. Although we were not using the open source media driver nor opensrc MSDK. The use case was to have the compositor sends the FB prime handle to be encoded by the gst-mfx plugins.
